rgw_sal_motr: [CORTX-33538] Fix list-objects api response

vastradparayya opened the pull request:
problem:
	list-objects API returning number of objects with
	Marker attibute is mismatching in s3tests test suite.
	bucket_name = _create_objects(keys=['foo', 'bar', 'baz'])
	client = get_client()
	response = client.list_objects(Bucket=bucket_name, Marker='baz', MaxKeys=2)
	Keys - ['baz', 'foo'] // We should be getting only foo here
	Below ceph s3tests are failing with assertion
	errors on setup with motr as backend. number of objects
	returning are mismatching with s3test.
	's3tests_boto3.functional.test_s3:test_bucket_list_many',
	's3tests_boto3.functional.test_s3:test_bucket_list_maxkeys_one',
	's3tests_boto3.functional.test_s3:test_bucket_listv2_many',
	's3tests_boto3.functional.test_s3:test_bucket_listv2_maxkeys_one'
solution:
	In list_objects(MotrBucket::list) of sal_motr code,
	modified the logic for listing objects with marker.
        Output for above s3tests is at the end of PR template

Signed-off-by: Parayya Vastrad <parayya.vastrad@seagate.com>





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".

  - The Signed-off-by line in every git commit is important; see <span class="x x-first x-last">[Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/master/</span>SubmittingPatches.rst<span class="x x-first x-last">)</span>
-->

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

(virtualenv) [root@ssc-vm-g2-rhev4-3360 s3-tests]# S3TEST_CONF=./ceph.conf virtualenv/bin/nosetests s3tests_boto3.functional.test_s3:test_bucket_list_many -v 2>&1 | tee -a cephtest.log
/root/s3-tests/virtualenv/lib/python3.6/site-packages/boto3/compat.py:88: PythonDeprecationWarning: Boto3 will no longer support Python 3.6 starting May 30, 2022. To continue receiving service updates, bug fixes, and security updates please upgrade to Python 3.7 or later. More information can be found here: https://aws.amazon.com/blogs/developer/python-support-policy-updates-for-aws-sdks-and-tools/
  warnings.warn(warning, PythonDeprecationWarning)
s3tests_boto3.functional.test_s3.test_bucket_list_many ... ok----------------------------------------------------------------------
Ran 1 test in 1.039sOK
(virtualenv) [root@ssc-vm-g2-rhev4-3360 s3-tests]# S3TEST_CONF=./ceph.conf virtualenv/bin/nosetests s3tests_boto3.functional.test_s3:test_bucket_list_maxkeys_one -v 2>&1 | tee -a cephtest.log
/root/s3-tests/virtualenv/lib/python3.6/site-packages/boto3/compat.py:88: PythonDeprecationWarning: Boto3 will no longer support Python 3.6 starting May 30, 2022. To continue receiving service updates, bug fixes, and security updates please upgrade to Python 3.7 or later. More information can be found here: https://aws.amazon.com/blogs/developer/python-support-policy-updates-for-aws-sdks-and-tools/
  warnings.warn(warning, PythonDeprecationWarning)
s3tests_boto3.functional.test_s3.test_bucket_list_maxkeys_one ... ok----------------------------------------------------------------------
Ran 1 test in 1.333sOK
(virtualenv) [root@ssc-vm-g2-rhev4-3360 s3-tests]# S3TEST_CONF=./ceph.conf virtualenv/bin/nosetests s3tests_boto3.functional.test_s3:test_bucket_listv2_many -v 2>&1 | tee -a cephtest.log
/root/s3-tests/virtualenv/lib/python3.6/site-packages/boto3/compat.py:88: PythonDeprecationWarning: Boto3 will no longer support Python 3.6 starting May 30, 2022. To continue receiving service updates, bug fixes, and security updates please upgrade to Python 3.7 or later. More information can be found here: https://aws.amazon.com/blogs/developer/python-support-policy-updates-for-aws-sdks-and-tools/
  warnings.warn(warning, PythonDeprecationWarning)
s3tests_boto3.functional.test_s3.test_bucket_listv2_many ... ok----------------------------------------------------------------------
Ran 1 test in 0.860sOK
(virtualenv) [root@ssc-vm-g2-rhev4-3360 s3-tests]# S3TEST_CONF=./ceph.conf virtualenv/bin/nosetests s3tests_boto3.functional.test_s3:test_bucket_listv2_maxkeys_one -v 2>&1 | tee -a cephtest.log
/root/s3-tests/virtualenv/lib/python3.6/site-packages/boto3/compat.py:88: PythonDeprecationWarning: Boto3 will no longer support Python 3.6 starting May 30, 2022. To continue receiving service updates, bug fixes, and security updates please upgrade to Python 3.7 or later. More information can be found here: https://aws.amazon.com/blogs/developer/python-support-policy-updates-for-aws-sdks-and-tools/
  warnings.warn(warning, PythonDeprecationWarning)
s3tests_boto3.functional.test_s3.test_bucket_listv2_maxkeys_one ... ok----------------------------------------------------------------------
Ran 1 test in 1.022sOK
(virtualenv) [root@ssc-vm-g2-rhev4-3360 s3-tests]#

sachinpunadikar reviewed (approved):
LGTM

sumedhak27 commented:
Hi @vastradparayya , the List function is a bit hard to understand can you please mention eg. of marker value for which the test was failing, and how the case is handled now.

sumedhak27 commented on src/rgw/rgw_sal_motr.cc:
```diff
@@ -1307,13 +1307,24 @@ int MotrBucket::list(const DoutPrefixProvider *dpp, ListParams& params, int max,
+                      (ent.key.instance != "" && 
+                      ent.key.instance < params.marker.instance))
+                    continue;
+              }else{
```
```suggestion
              } else {
```

vastradparayya replied:
Done

vastradparayya commented:
> Hi @vastradparayya, the List function is a bit hard to understand can you please mention eg. of marker value for which the test was failing, and how the case is handled now. Also, the PR title or description should not say that we are fixing some failed test, it should say what issue we are solving.



> Hi @vastradparayya, the List function is a bit hard to understand can you please mention eg. of marker value for which the test was failing, and how the case is handled now. Also, the PR title or description should not say that we are fixing some failed test, it should say what issue we are solving.

response = client.list_objects(Bucket=bucket_name, **Marker='baz',**MaxKeys=2)
**MGW**
Keys -  ['bar', 'baz']
Keys after truncate -  ['bar', 'baz']
**Keys -  ['baz', 'foo']** // eq(len(keys), 1) // Here we are getting assert error


response = client.list_objects(Bucket=bucket_name, **Marker='baz'**,MaxKeys=2)
**ceph**
Keys -  ['bar', 'baz']
Keys after truncate -  ['bar', 'baz']
**Keys -  ['foo']**

sumedhak27 commented:
> > Hi @vastradparayya, the List function is a bit hard to understand can you please mention eg. of marker value for which the test was failing, and how the case is handled now. Also, the PR title or description should not say that we are fixing some failed test, it should say what issue we are solving.
> 
> > Hi @vastradparayya, the List function is a bit hard to understand can you please mention eg. of marker value for which the test was failing, and how the case is handled now. Also, the PR title or description should not say that we are fixing some failed test, it should say what issue we are solving.
> 
> response = client.list_objects(Bucket=bucket_name, **Marker='baz',**MaxKeys=2) **MGW** Keys - ['bar', 'baz'] Keys after truncate - ['bar', 'baz'] **Keys - ['baz', 'foo']** // eq(len(keys), 1) // Here we are getting assert error
> 
> response = client.list_objects(Bucket=bucket_name, **Marker='baz'**,MaxKeys=2) **ceph** Keys - ['bar', 'baz'] Keys after truncate - ['bar', 'baz'] **Keys - ['foo']**

can you pls verify the output for the following case.
1. keys = {foo/bar1, foo/bar2,  hoo/bar1, hoo/bar2}
2.  list bucket, delimeter = "/" ,  marker = "foo/"
3.  expected output is {hoo/}

sumedhak27 commented on src/rgw/rgw_sal_motr.cc:
```diff
@@ -1307,13 +1307,24 @@ int MotrBucket::list(const DoutPrefixProvider *dpp, ListParams& params, int max,
               // filter out versions which go before marker.instance
-              ((!null_ent.key.empty() && params.marker.instance == "null" &&
+              // null_ent is not yet populated
+              if (null_ent.key.name == "") {
+                 if ((params.marker.instance == "" && 
+                      ent.key.instance == "") || 
+                      (ent.key.instance != "" && 
+                      ent.key.instance < params.marker.instance))
```
I see some handling was done previously for `params.marker.instance == "null"` and that condition is not checked now. 
so testing of `list-object-versions` api is also necessary if not done already.

vastradparayya replied:
Hi @sumedhak27 when we check the absence value for `params.marker.instance` is `""`.
For `params.marker.instance == "null"` condition was not satisfying

sjain09 replied:
Since this API logic is bit tricky and the changes overlaps with PR #349, I would suggest to first merge PR #349 and rebase this PR and create a custom build to test all the possible cases with list api before merging into main.

vastradparayya replied:
Done. Rebased with main (#349) changes

sumedhak27 replied:
> I see some handling was done previously for params.marker.instance == "null" and that condition is not checked now.
so testing of list-object-versions api is also necessary if not done already.

Is this tested now @vastradparayya? I only have concern about this `null` version handling now.

vastradparayya commented:
> > > Hi @vastradparayya, the List function is a bit hard to understand can you please mention eg. of marker value for which the test was failing, and how the case is handled now. Also, the PR title or description should not say that we are fixing some failed test, it should say what issue we are solving.
> > 
> > 
> > response = client.list_objects(Bucket=bucket_name, **Marker='baz',**MaxKeys=2) **MGW** Keys - ['bar', 'baz'] Keys after truncate - ['bar', 'baz'] **Keys - ['baz', 'foo']** // eq(len(keys), 1) // Here we are getting assert error
> > response = client.list_objects(Bucket=bucket_name, **Marker='baz'**,MaxKeys=2) **ceph** Keys - ['bar', 'baz'] Keys after truncate - ['bar', 'baz'] **Keys - ['foo']**
> 
> can you pls verify the output for the following case.
> 
> 1. keys = {foo/bar1, foo/bar2,  hoo/bar1, hoo/bar2}
> 2. list bucket, delimeter = "/" ,  marker = "foo/"
> 3. expected output is {hoo/}

Hi @sumedhak27 we are getting error when we try above scenario with plain ceph. 
Same error getting with motr as backend
`AssertionError: 0 != 1`

vastradparayya commented:
> > > > Hi @vastradparayya, the List function is a bit hard to understand can you please mention eg. of marker value for which the test was failing, and how the case is handled now. Also, the PR title or description should not say that we are fixing some failed test, it should say what issue we are solving.
> > > 
> > > 
> > > response = client.list_objects(Bucket=bucket_name, **Marker='baz',**MaxKeys=2) **MGW** Keys - ['bar', 'baz'] Keys after truncate - ['bar', 'baz'] **Keys - ['baz', 'foo']** // eq(len(keys), 1) // Here we are getting assert error
> > > response = client.list_objects(Bucket=bucket_name, **Marker='baz'**,MaxKeys=2) **ceph** Keys - ['bar', 'baz'] Keys after truncate - ['bar', 'baz'] **Keys - ['foo']**
> > 
> > 
> > can you pls verify the output for the following case.
> > 
> > 1. keys = {foo/bar1, foo/bar2,  hoo/bar1, hoo/bar2}
> > 2. list bucket, delimeter = "/" ,  marker = "foo/"
> > 3. expected output is {hoo/}
> 
> Hi @sumedhak27 we are getting error when we try above scenario with plain ceph. Same error getting with motr as backend `AssertionError: 0 != 1`

Yes, @sumedhak27, I tested with s3test, as below - 
bucket_name = _create_objects(keys=['foo/bar1', 'foo/bar2', 'hoo/bar1', 'hoo/bar2'])
client = get_client()
response = client.list_objects(Bucket=bucket_name, Delimiter = "/", Marker = "foo/",MaxKeys=2)
// output
Response -  {'ResponseMetadata': {'HTTPStatusCode': 200, 'HTTPHeaders': {'transfer-encoding': 'chunked', 'content-type': 'application/xml', 'date': 'Tue, 19 Jul 2022 13:44:23 GMT', 'connection': 'Keep-Alive'}, 'RetryAttempts': 0}, 'IsTruncated': False, 'Marker': 'foo/', 'Name': 'yournamehere-zo6ha4d1jm97ar8a-1', 'Prefix': '', 'Delimiter': '/', 'MaxKeys': 2, 'CommonPrefixes': [**{'Prefix': 'hoo/'}**], 'EncodingType': 'url'}

sumedhak27 commented:
Only list-object versions is yet to be tested now r8 @vastradparayya?

sjain09 commented:
@vastradparayya please update the problem in the description with actual issue/bug in w.r.t. list-objects api. List all the use cases which were failing earlier. Then list the ceph-s3 test cases that are failing due to the bug.

priyanka25081999 commented:
Hi @vastradparayya 
Could you please test the below scenario and confirm if the null version object entry is present in the List object version API response ?
```
1. Create a bucket (test-bucket1) and enable the bucket versioning
2. Put an object (object1) - 2 times 
3. Suspend the bucket versioning
4. Put an object (object1)
5. List object versions with --key-marker object1
```

vastradparayya commented:
> 

Hi @sjain09 I have rebased with main and updated problem description in PR template. And these are the s3tests related to bucket list are failing. 
@kapiljinna please confirm.

vastradparayya commented:
> 

Hi @priyanka25081999, output for above scenario
`[root@ssc-vm-rhev4-2390 build]# aws s3api create-bucket --bucket test-bucket1 --profile user1 --endpoint $endpoint
[root@ssc-vm-rhev4-2390 build]# aws s3api put-bucket-versioning --bucket test-bucket1 --versioning-configuration Status=Enabled --profile user1 --endpoint $endpoint
[root@ssc-vm-rhev4-2390 build]# aws s3api put-object --bucket test-bucket1 --key object1 --body ceph.conf --profile user1 --endpoint $endpoint
{
    "ETag": "\"bc546ab3d89b45cafed2780db5c43a5d\"",
    "VersionId": "zWnEToGMHoa73o7jEINdMlpUD1K772h"
}
[root@ssc-vm-rhev4-2390 build]# aws s3api put-object --bucket test-bucket1 --key object1 --body ceph.conf --profile user1 --endpoint $endpoint
{
    "ETag": "\"bc546ab3d89b45cafed2780db5c43a5d\"",
    "VersionId": "zWnETnQsG2eavgBMUMnOF4ZNosBAgU3"
}
[root@ssc-vm-rhev4-2390 build]# aws s3api put-bucket-versioning --bucket test-bucket1 --versioning-configuration Status=Suspended --profile user1 --endpoint $endpoint
[root@ssc-vm-rhev4-2390 build]# aws s3api put-object --bucket test-bucket1 --key object1 --body ceph.conf --profile user1 --endpoint $endpoint                       {
    "ETag": "\"bc546ab3d89b45cafed2780db5c43a5d\""
}
[root@ssc-vm-rhev4-2390 build]# aws s3api list-object-versions --bucket test-bucket1 --key-marker object1 --profile user1 --endpoint $endpoint
{
    "IsTruncated": false,
    "KeyMarker": "object1",
    "VersionIdMarker": "",
    "Versions": [
        {
            "ETag": "\"bc546ab3d89b45cafed2780db5c43a5d\"",
            "Size": 6144,
            "StorageClass": "STANDARD",
            "Key": "object1",
            "VersionId": "zWnETnQsG2eavgBMUMnOF4ZNosBAgU3",
            "IsLatest": false,
            "LastModified": "2022-07-27T06:16:28.626Z",
            "Owner": {
                "DisplayName": "user1",
                "ID": "user1"
            }
        },
        {
            "ETag": "\"bc546ab3d89b45cafed2780db5c43a5d\"",
            "Size": 6144,
            "StorageClass": "STANDARD",
            "Key": "object1",
            "VersionId": "zWnEToGMHoa73o7jEINdMlpUD1K772h",
            "IsLatest": false,
            "LastModified": "2022-07-27T06:16:25.691Z",
            "Owner": {
                "DisplayName": "user1",
                "ID": "user1"
            }
        }
    ],
    "Name": "test-bucket1",
    "Prefix": "",
    "MaxKeys": 1000,
    "EncodingType": "url"
}
[root@ssc-vm-rhev4-2390 build]#`

vastradparayya commented:
> Hi @vastradparayya Could you please test the below scenario and confirm if the null version object entry is present in the List object version API response ?
> 
> ```
> 1. Create a bucket (test-bucket1) and enable the bucket versioning
> 2. Put an object (object1) - 2 times 
> 3. Suspend the bucket versioning
> 4. Put an object (object1)
> 5. List object versions with --key-marker object1
> ```

Hi @priyanka25081999 , output for above scenario. please confirm, is this correct behavior

[root@ssc-vm-rhev4-2390 build]# aws s3api create-bucket --bucket test-bucket1 --profile user1 --endpoint $endpoint
[root@ssc-vm-rhev4-2390 build]# aws s3api put-bucket-versioning --bucket test-bucket1 --versioning-configuration Status=Enabled --profile user1 --endpoint $endpoint
[root@ssc-vm-rhev4-2390 build]# aws s3api put-object --bucket test-bucket1 --key object1 --body ceph.conf --profile user1 --endpoint $endpoint
{
    "ETag": "\"bc546ab3d89b45cafed2780db5c43a5d\"",
    "VersionId": "zWnEToGMHoa73o7jEINdMlpUD1K772h"
}
[root@ssc-vm-rhev4-2390 build]# aws s3api put-object --bucket test-bucket1 --key object1 --body ceph.conf --profile user1 --endpoint $endpoint
{
    "ETag": "\"bc546ab3d89b45cafed2780db5c43a5d\"",
    "VersionId": "zWnETnQsG2eavgBMUMnOF4ZNosBAgU3"
}
[root@ssc-vm-rhev4-2390 build]# aws s3api put-bucket-versioning --bucket test-bucket1 --versioning-configuration Status=Suspended --profile user1 --endpoint $endpoint
[root@ssc-vm-rhev4-2390 build]# aws s3api put-object --bucket test-bucket1 --key object1 --body ceph.conf --profile user1 --endpoint $endpoint                       {
    "ETag": "\"bc546ab3d89b45cafed2780db5c43a5d\""
}
[root@ssc-vm-rhev4-2390 build]# aws s3api list-object-versions --bucket test-bucket1 --key-marker object1 --profile user1 --endpoint $endpoint
{
    "IsTruncated": false,
    "KeyMarker": "object1",
    "VersionIdMarker": "",
    "Versions": [
        {
            "ETag": "\"bc546ab3d89b45cafed2780db5c43a5d\"",
            "Size": 6144,
            "StorageClass": "STANDARD",
            "Key": "object1",
            "VersionId": "zWnETnQsG2eavgBMUMnOF4ZNosBAgU3",
            "IsLatest": false,
            "LastModified": "2022-07-27T06:16:28.626Z",
            "Owner": {
                "DisplayName": "user1",
                "ID": "user1"
            }
        },
        {
            "ETag": "\"bc546ab3d89b45cafed2780db5c43a5d\"",
            "Size": 6144,
            "StorageClass": "STANDARD",
            "Key": "object1",
            "VersionId": "zWnEToGMHoa73o7jEINdMlpUD1K772h",
            "IsLatest": false,
            "LastModified": "2022-07-27T06:16:25.691Z",
            "Owner": {
                "DisplayName": "user1",
                "ID": "user1"
            }
        }
    ],
    "Name": "test-bucket1",
    "Prefix": "",
    "MaxKeys": 1000,
    "EncodingType": "url"
}
[root@ssc-vm-rhev4-2390 build]#

priyanka25081999 commented:
> > Hi @vastradparayya Could you please test the below scenario and confirm if the null version object entry is present in the List object version API response ?
> > ```
> > 1. Create a bucket (test-bucket1) and enable the bucket versioning
> > 2. Put an object (object1) - 2 times 
> > 3. Suspend the bucket versioning
> > 4. Put an object (object1)
> > 5. List object versions with --key-marker object1
> > ```
> 
> Hi @priyanka25081999 , output for above scenario. please confirm, is this correct behavior
> 
> [root@ssc-vm-rhev4-2390 build]# aws s3api create-bucket --bucket test-bucket1 --profile user1 --endpoint $endpoint [root@ssc-vm-rhev4-2390 build]# aws s3api put-bucket-versioning --bucket test-bucket1 --versioning-configuration Status=Enabled --profile user1 --endpoint $endpoint [root@ssc-vm-rhev4-2390 build]# aws s3api put-object --bucket test-bucket1 --key object1 --body ceph.conf --profile user1 --endpoint $endpoint { "ETag": ""bc546ab3d89b45cafed2780db5c43a5d"", "VersionId": "zWnEToGMHoa73o7jEINdMlpUD1K772h" } [root@ssc-vm-rhev4-2390 build]# aws s3api put-object --bucket test-bucket1 --key object1 --body ceph.conf --profile user1 --endpoint $endpoint { "ETag": ""bc546ab3d89b45cafed2780db5c43a5d"", "VersionId": "zWnETnQsG2eavgBMUMnOF4ZNosBAgU3" } [root@ssc-vm-rhev4-2390 build]# aws s3api put-bucket-versioning --bucket test-bucket1 --versioning-configuration Status=Suspended --profile user1 --endpoint $endpoint [root@ssc-vm-rhev4-2390 build]# aws s3api put-object --bucket test-bucket1 --key object1 --body ceph.conf --profile user1 --endpoint $endpoint { "ETag": ""bc546ab3d89b45cafed2780db5c43a5d"" } [root@ssc-vm-rhev4-2390 build]# aws s3api list-object-versions --bucket test-bucket1 --key-marker object1 --profile user1 --endpoint $endpoint { "IsTruncated": false, "KeyMarker": "object1", "VersionIdMarker": "", "Versions": [ { "ETag": ""bc546ab3d89b45cafed2780db5c43a5d"", "Size": 6144, "StorageClass": "STANDARD", "Key": "object1", "VersionId": "zWnETnQsG2eavgBMUMnOF4ZNosBAgU3", "IsLatest": false, "LastModified": "2022-07-27T06:16:28.626Z", "Owner": { "DisplayName": "user1", "ID": "user1" } }, { "ETag": ""bc546ab3d89b45cafed2780db5c43a5d"", "Size": 6144, "StorageClass": "STANDARD", "Key": "object1", "VersionId": "zWnEToGMHoa73o7jEINdMlpUD1K772h", "IsLatest": false, "LastModified": "2022-07-27T06:16:25.691Z", "Owner": { "DisplayName": "user1", "ID": "user1" } } ], "Name": "test-bucket1", "Prefix": "", "MaxKeys": 1000, "EncodingType": "url" } [root@ssc-vm-rhev4-2390 build]#

@vastradparayya In the above list-object-versions response, we should get null object version entry of object1.
With the main branch code, this scenario is working as expected.

priyanka25081999 commented on src/rgw/rgw_sal_motr.cc:
```diff
@@ -1328,13 +1328,24 @@ int MotrBucket::list(const DoutPrefixProvider *dpp, ListParams& params, int max,
+                    continue;
+              } else {
+              // null_ent is populated
+              if ((params.marker.instance == "" &&
```
@vastradparayya could you please explain why this check is required ?

vastradparayya replied:
@priyanka25081999 we separated null_ent before populating and after. First if is for first time request where null_ent is not populated.

sjain09 commented:
@vastradparayya  please resolve the merge conflicts

vastradparayya commented:
Reopening

sachinpunadikar reviewed (approved):
LGTM

sachinpunadikar reviewed (approved):
+1

sachinpunadikar commented:
The required code changes are handled by PR https://github.com/Seagate/cortx-rgw/pull/374
